app/models/pipeline_run.rb

```diff
@@ -153,38 +157,255 @@ def retry
+      # postprocessing succeeded
+      "COMPLETE"
+    elsif h["db_load_alignment"] == STATUS_LOADED
+      # alignment succeeded, postprocessing in progress
+      "POST PROCESSING"
+    elsif h["db_load_host_filtering"] == STATUS_LOADED
+      # host-filtering succeeded, alignment in progress
+      "ALIGNMENT"
+    else
+      # host-filtering in progress
+      "HOST FILTERING"
+    end
+  end
+
+  def status_display_pre_result_monitor(run_stages)
+    status_by_loader = {}
+    run_stages.each do |rs|
+      status_by_loader[rs.load_db_command_func] = rs.job_status
+    end
+    status_display(status_by_loader)
+  end
+
+  def status_display_pre_run_stages
+    if %w[CHECKED SUCCEEDED].include?(job_status)
+      'COMPLETE'
+    elsif %w[FAILED ERROR].include?(job_status)
+      'FAILED'
+    elsif %w[RUNNING LOADED].include?(job_status)
+      'IN PROGRESS'
+    else
+      'WAITING'
+    end
+  end
+
+  def check_and_enqueue(db_load_command_name)
+    # TODO: handle case where resque crashes and needs to be restarted. What happens to runs that were queued to load results?
+    if output_ready?(db_load_command_name) && ![STATUS_LOADED, STATUS_LOADING].include?(result_status_for(db_load_command_name))
+      update_result_status(db_load_command_name, STATUS_LOADING)
+      Resque.enqueue(ResultMonitorLoader, id, db_load_command_name)
+    end
+  end
+
+  def monitor_results
+    return if results_finalized?
+
+    # When last output has just been loaded, indicate sample completion via results_finalized:
+    if result_status_for("db_load_postprocess") == STATUS_LOADED # last output has been loaded
+      self.results_finalized = 1 # this ensures in_progress will be false and monitor_results won't be run again on this pipeline run
       save
-      if sample.project.complete?
+      if sample.project.results_complete? # the entire project has just completed
         notify_users
         sample.project.create_or_update_project_background if sample.project.background_flag == 1
       end
+      return
+    end
+
+    # Otherwise go through pipeline result files and load if available.
+
+    # First, get pipeline_version, which will determine S3 locations of output files:
+    update(pipeline_version: fetch_pipeline_version) if pipeline_version.blank?
+
+    # Then update the job stats, which the pipeline can modify incrementally:
+    # [TODO: make it a single file rather than using stage-specific.]
+    host_filtering_job_stats_s3 = "#{host_filter_output_s3_path}/#{STATS_JSON_NAME}"
+    alignment_job_stats_s3 = "#{alignment_output_s3_path}/#{STATS_JSON_NAME}"
+    if file_generated_since_jobstats?(alignment_job_stats_s3)
+      load_job_stats(alignment_job_stats_s3)
+    elsif file_generated_since_jobstats?(host_filtering_job_stats_s3)
+      load_job_stats(host_filtering_job_stats_s3)
+    end
+
+    # Finally, load any new outputs that have become available:
+    check_and_enqueue("db_load_host_filtering")
```

Comment:
instead of doing this, I think you can make it true result_monitor by 
load_version_file() unless pipeline_version # blocking call  
check_and_enqueue("ercc") 
check_and_enqueue("taxon_counts") 
check_and_enqueue("taxon_byte_ranges") .... 
load_stats_file
and set finalized=1 when all the required files are generated

Reply:
Ok! I will make this change today